client/src/context/SocketContext.js

```diff
@@ -0,0 +1,50 @@
+import React, { createContext, useState, useEffect } from 'react'
+import io from 'socket.io-client'
+
+export const SocketContext = createContext()
+
+let newSocket;
+const CONNECTION_PORT = 'localhost:3001'
+
+const SocketProvider = ({ children }) => {
+    const [currentUsers, setCurrentUsers] = useState(0)
+    const [socket, setSocket] = useState(null)
+
+
+    const setupSocket = () => {
+        if (!socket) {
+            newSocket = io(CONNECTION_PORT)
+
+            newSocket.on('connect', () => {
+                console.log('socket connected!')
+            })
+
+            newSocket.on('disconnect', () => {
+                setSocket(null)
+                setTimeout(setupSocket, 3000)
+                console.log('socket disconnected!')
+            })
+            setSocket(newSocket)
+        }
+    }
+
+    useEffect(() => {
+        if (socket) {
+            socket.on('user count', data => {
+                setCurrentUsers(data)
+            })
+        }
+    })
```

Comment:
I don't think we need this in a useEffect. The useEffect without a dependency array is gonna get called constantly.
Checkout [this tutorial](https://alexboots.medium.com/using-react-context-with-socket-io-3b7205c86a6d) to get an idea for an approach here